pkg/dashboard/manager.go

```diff
@@ -0,0 +1,230 @@
+	return &Manager{
+		ctx:        ctx,
+		cancel:     cancel,
+		srv:        srv,
+		service:    s,
+		redirector: redirector,
+	}
+}
+
+func (m *Manager) start() {
+	m.wg.Add(1)
+	go func() {
+		defer logutil.LogPanic()
+		defer m.wg.Done()
+
+		status := unknown
+		for status != stopped {
+			status = m.serviceLoop(status)
+		}
+	}()
+}
+
+func (m *Manager) stop() {
+	m.cancel()
+	m.wg.Wait()
+	log.Info("exit dashboard loop")
+}
+
+func (m *Manager) serviceLoop(lastStatus electionStatus) (nextStatus electionStatus) {
+	var ticker *time.Ticker
+	switch lastStatus {
+	case unknown:
+		ticker = time.NewTicker(FastCheckInterval)
+	case known:
+		ticker = time.NewTicker(SlowCheckInterval)
+	default:
+		panic("unreachable")
+	}
+	defer ticker.Stop()
+
+	nextStatus = lastStatus
+	for nextStatus == lastStatus {
+		select {
+		case <-m.ctx.Done():
+			m.stopService()
+			return stopped
+		case <-ticker.C:
+			nextStatus = m.checkAddress()
+		}
+	}
+
+	return
+}
+
+// checkDashboardAddress checks if the dashboard service needs to change due to dashboard address is changed.
+func (m *Manager) checkAddress() electionStatus {
+	dashboardAddress := m.srv.GetScheduleOption().GetDashboardAddress()
+	switch dashboardAddress {
+	case "auto":
+		if m.srv.GetMember().IsLeader() {
+			m.setNewDashboardAddress()
+		}
+		return unknown
+	case "none":
+		m.redirector.SetAddress("")
+		m.stopService()
+		return known
+	default:
+		if _, err := url.Parse(dashboardAddress); err != nil {
+			log.Error("illegal dashboard address", zap.String("address", dashboardAddress))
+			return unknown
+		}
+		if m.srv.GetMember().IsLeader() && !m.isOnline(dashboardAddress) {
+			m.setNewDashboardAddress()
+			return unknown
+		}
+	}
+
+	m.redirector.SetAddress(dashboardAddress)
+
+	clientUrls := m.srv.GetMemberInfo().GetClientUrls()
+	if len(clientUrls) > 0 && clientUrls[0] == dashboardAddress {
+		m.startService()
+	} else {
+		m.stopService()
+	}
+
+	return known
+}
+
+func (m *Manager) isOnline(addr string) bool {
+	rc := m.srv.GetRaftCluster()
+	if rc == nil || !rc.IsRunning() {
+		return true
+	}
+	members, err := cluster.GetMembers(m.srv.GetClient())
+	if err != nil {
+		log.Error("failed to get members")
+		return true
+	}
+
+	for _, member := range members {
+		if member.GetClientUrls()[0] == addr {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (m *Manager) setNewDashboardAddress() {
+	rc := m.srv.GetRaftCluster()
+	if rc == nil || !rc.IsRunning() {
+		return
+	}
+	members, err := cluster.GetMembers(m.srv.GetClient())
+	if err != nil || len(members) == 0 {
+		log.Error("failed to get members")
+		return
+	}
+
+	// get new dashboard address
+	var addr string
+	switch len(members) {
+	case 0:
+		return
+	case 1:
+		addr = members[0].GetClientUrls()[0]
+	default:
+		addr = members[0].GetClientUrls()[0]
+		leaderID := m.srv.GetMemberInfo().MemberId
+		sort.Slice(members, func(i, j int) bool { return members[i].GetMemberId() < members[j].GetMemberId() })
```

Comment:
why do we need to sort the member id?

Reply:
We want to select the follower node with the smallest member id.